llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -157,6 +159,55 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
   return true;
 }
 
+bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  unsigned DstRegLen = MRI.getType(DstReg).getElementCount().getKnownMinValue();
+  SmallVector<Register, 16> Regs;
+  MachineInstr *CurMI = &MI;
+  int NumInserts = DstRegLen;
+
+  auto IsSet = [&](const MachineInstr *MI) {
+    return MI->getOpcode() == AIE2::G_INTRINSIC &&
+           cast<GIntrinsic>(*MI).getIntrinsicID() ==
+               Intrinsic::aie2_set_I512_I128;
+  };
+
+  auto IsVInsert = [&](const MachineInstr *MI) {
+    return MI->getOpcode() == AIE2::G_INTRINSIC &&
+           cast<GIntrinsic>(*MI).getIntrinsicID() ==
+               Intrinsic::aie2_vinsert8_I512;
+  };
+
+  while (NumInserts--) {
+    const Register SrcReg = CurMI->getOperand(2).getReg();
+    const Register IdxReg = CurMI->getOperand(3).getReg();
+    const Register SclSrcReg = CurMI->getOperand(4).getReg();
+
+    auto Cst = getIConstantVRegValWithLookThrough(IdxReg, MRI);
+    if (!Cst || Cst->Value.getSExtValue() != NumInserts)
+      return false;
+    Regs.push_back(SclSrcReg);
+
+    MachineInstr *SrcMI = MRI.getUniqueVRegDef(SrcReg);
+    assert(SrcMI && "Expected SSA.");
+    if (IsSet(SrcMI) && !tryToCombineSetExtract(*SrcMI))
+      break;
+
+    SrcMI = getDefIgnoringCopies(SrcReg, MRI);
+    if (!IsVInsert(SrcMI))
+      break;
+    CurMI = SrcMI;
+  }
+  if (!++NumInserts)
```

Comment:
This is not a very clear condition. I think it corresponds to the break from the loop above. Why not return false immediately instead of breaking?